pandas/io/excel/_xlsxwriter.py

```diff
@@ -271,14 +277,35 @@ def _write_cells(
                 style = self.book.add_format(_XlsxStyler.convert(cell.style, fmt))
                 style_dict[stylekey] = style
 
+            abs_row = startrow + cell.row
+            abs_col = startcol + cell.col
+
+            # track bounds
+            if min_r is None or abs_row < min_r:
+                min_r = abs_row
+            if min_c is None or abs_col < min_c:
+                min_c = abs_col
+            if max_r is None or abs_row > max_r:
+                max_r = abs_row
+            if max_c is None or abs_col > max_c:
+                max_c = abs_col
+
             if cell.mergestart is not None and cell.mergeend is not None:
                 wks.merge_range(
-                    startrow + cell.row,
-                    startcol + cell.col,
+                    abs_row,
+                    abs_col,
                     startrow + cell.mergestart,
                     startcol + cell.mergeend,
                     val,
                     style,
                 )
             else:
-                wks.write(startrow + cell.row, startcol + cell.col, val, style)
+                wks.write(abs_row, abs_col, val, style)
+
+        if autofilter and min_r is not None and min_c is not None and max_r is not None and max_c is not None:
+            # Apply autofilter over the used range. xlsxwriter uses 0-based indices.
+            try:
+                wks.autofilter(min_r, min_c, max_r, max_c)
+            except Exception:
+                # Be resilient if engine version doesn't support or range invalid
+                pass
```

Comment:
How could the engine version not support it? How could the range be invalid? Catching all errors is a great way to hide bugs, we should avoid doing so if at all possible.